compiler/transformation/transform.rs

```diff
@@ -31,3 +41,29 @@ pub fn apply_transformations(
     // - function inlining v2: we could try to inline/lift some constraints from recursive calls into the parent query to dramatically cut the search space
     // - function inlining v3: we could introduce new sub-patterns that include sort/offset/limit that let us more generally inline functions?
 }
+
+fn prune_redundant_roleplayer_deduplication(conjunction: &mut Conjunction, block_annotations: &mut TypeAnnotations) {
+    // TODO: If either the role-players or the players
+    conjunction.constraints_mut().constraints_mut().retain(|constraint| {
+        if let Constraint::RolePlayerDeduplication(dedup) = constraint {
+            let first = block_annotations
+                .constraint_annotations_of(Constraint::Links(dedup.links1().clone()))
+                .unwrap()
+                .as_links()
+                .player_to_role();
+            let second = block_annotations
+                .constraint_annotations_of(Constraint::Links(dedup.links2().clone()))
+                .unwrap()
+                .as_links()
+                .player_to_role();
+            first.iter().any(|(player, role_types)| {
+                return second
+                    .get(&player)
+                    .map(|type_set| role_types.iter().any(|role_type| type_set.contains(role_type)))
+                    .unwrap_or(false);
+            })
+        } else {
+            true
+        }
```

Comment:
If the first and second Links constraints have no `(RoleType, PlayerType)` pair in common, they will never use the same edge and can be safely pruned.